ops/ccs-ops-misc/load_test/common/bfd_user_base.py

```diff
@@ -0,0 +1,155 @@
+        # Report the various response time percentiles against the SLA
+        if hasattr(self, 'SLA_BASELINE') and self.SLA_BASELINE:
+            validation.check_sla_validation(self.environment, self.SLA_BASELINE)
+
+
+    def get_by_url(self, url: str, headers: Dict[str, str] = None,
+            name: str = ''):
+        '''Send one GET request and parse the response for pagination.
+
+        This method extends Locust's HttpUser::client.get() method to make
+        creating the requests nicer. Specifically, the query string parameters
+        are specified as a separate dictionary opposed to part of the path, the
+        cert and verify arguments (which will never change) are already set,
+        and Cache-Control headers are automatically set to ensure caching is
+        disabled.
+        '''
+
+        safe_headers = {} if headers is None else headers
+
+        with self.client.get(url,
+                cert=self.client_cert,
+                verify=self.server_public_key,
+                headers={**safe_headers, 'Cache-Control': 'no-store, no-cache'},
+                name=name,
+                catch_response=True
+        ) as response:
+            # This is normally where you would validate a response, but we
+            # don't need to do that. Instead, we're checking it for valid
+            # "next" URLs that we can add to a URL pool.
+            next_url = BFDUserBase.get_next_url(response.text)
+            if next_url is not None:
+                if name not in self.url_pools:
+                    self.url_pools[name] = []
+                self.url_pools[name].append(next_url)
+
+
+    def run_task(self, url_callback: Callable, headers: Dict[str, str] = None,
+            name: str = ''):
+        '''Figure out which URL we should query next and query the server'''
+
+        # First, see if we can generate a URL using the callback
+        try:
+            url = url_callback(self)
+        except IndexError:
+            url = None
+
+        if url is None and name in self.url_pools and self.url_pools[name]:
+            # We can't generate a URL from the callback, but there are still
+            # URLs in the pool of paginated data. Use that.
+            # url = self.url_pools[name].pop()
+            pass
+
+        if url is not None:
+            # Run the test using the URL we found
+            self.get_by_url(url=url, headers=headers, name=name)
+        # If no URL is found, then this test isn't counted in statistics
+
+
+    # Helper Functions
+
+
+    def load_data(self, flag_name: str, load_function: Callable, *args) -> List:
+        '''If the given DATA_REQUIRED flag exists on this class, pre-load the
+        data from the database.'''
+        if hasattr(self, 'DATA_REQUIRED') and flag_name in self.DATA_REQUIRED:
+            data_list = data.load_data_segment(load_function, *args).copy()
+            random.shuffle(data_list)
+            return data_list
+        else:
+            return []
```

Comment:
I _think_ there's a difference in behavior with how these changes load the data from the database with respect to how the current code does it. Before, the code would load the data _once_ from the database regardless of how many concurrent Locust `User`s there were, whereas with these changes it seems that each Locust `User` would independently query the database to get its data.

I'm not sure this is something we should be concerned with, _especially_ if we start using `TABLESAMPLE` and want to make sure our input data isn't being cached as much as possible, though.

Reply:
Indeed, with the distributed mode it would load once and split the data amongst the users. The data gathering part of the tests is expensive (in terms of both database weight and time) so we'd ideally not want to be calling for data multiple times, especially when the early users may have started hitting the service as we're still querying data. it would be ideal to gather the data once, if indeed we are doing it multiple times here.

Reply:
I moved the loading of those IDs and the code that relies on them into children of BFDUserBase. It should greatly reduce the amount of test duplication also -- each test should now be defined in a single place.

This sets us up for making the tests run based solely on Locust tags, but I'll leave that for a future item if we decide to do it.

Reply:
I've checked pretty thoroughly, and I'm pretty sure every worker thread is loading up its own separate database query. With `TABLESAMPLE` enabled, it's almost guaranteed to be totally different results. Personally, I think that's sufficient, and my preference is to rely on that for the `bene_id` and `hashed_mbi`. I'll take out the part that splits up the results between workers, because we won't even have the same results, so why discard perfectly useful `bene_id`s?

It's a little different for the contracts -- there are only a few, so we end up with the same data in each thread. They're randomized, but I think we'll still hit the same endpoints multiple times. For these, it probably is worthwhile to divide the results between the workers.

I'm definitely open to suggestions. I think if we want to truly guarantee separate datasets for the different threads, then we'd need to implement something with messaging between the master and worker threads. But I don't think anything like that is necessary for the testing we're doing so far.